src/routes/lobby/LobbyView.vue

```diff
@@ -121,7 +119,20 @@ export default {
   setup() {
     const { t } = useI18n();
     const gameStarted = ref(false);
-
+    const joinAudio = new Audio('/sounds/lobby/enter-lobby.mp3');
+    const leaveAudio = new Audio('/sounds/lobby/leave-lobby.mp3');
+    onMounted(() => {
+      setTimeout(() => {
+        if (joinAudio.readyState === 4) {
+          joinAudio.play();
+        }
+      }, 500);
```

Comment:
yes when page is mounted it does not have audio to play immediately so i use setTimeout.

Reply:
The issue with setTimeout is that it's a fixed delay, but the file may take more or less time than that to be ready.

I've been doing some more digging, and I think that what we should is to register an event listener to play the sound when it's loaded. There's an event called [canplaythrough](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/canplaythrough_event) that fires when the media is loaded. So maybe what we can do here is define a function to use event listeners to play the sound when it's ready. Something like this

```
function playAudioWhenReady(audio) {
  if (joinAudio.readyState === 4) {
    audio.play();
  } else {
    audio.removeEventListener('canplaythrough', audio.play);
    audio.addEventListener('canplaythrough', audio.play);
  }
}

onMounted(() => {
  playAudioWhenReady(joinAudio);
});

onUnmounted(() => {
  playAudioWhenReady(leaveAudio);
});
```

Then similarly down below we can call this function in the watcher to simplify that call and also ensure that things only play when loaded.

```
  watch: {
    opponentUsername(newVal) {
      if (newVal) {
        playAudioWhenReady(joinAudio);
      } else {
        playAudioWhenReady(leaveAudio);
      }
    }
  }
```

Reply:
Going to mirror what @itsalaidbacklife said here, setTimeout is a hard no because it's a fixed delay for a dynamic problem. Why not make a composable, so it's reusable for any sounds we want to use in the app? Something like this:

https://stackblitz.com/edit/vitejs-vite-mb5u4c?file=src%2FApp.vue,src%2Fcomposables%2Fuse-audio.js&terminal=dev

```js
import { ref } from 'vue';

export default function useAudio(url) {
  const audio = ref(null);

  const playAudio = () => {
    // Always play the audio file first it we have it
    if (audio.value) {
      audio.value.play();
      return;
    }

    audio.value = new Audio(url);
    console.debug('Initializing audio', {
      url,
      audio,
    });

    audio.value.addEventListener('canplaythrough', () => {
      try {
        audio.value.play();
      } catch (err) {
        console.warn('Unable to autoplay audio', {
          url,
          err,
        });
      }
    });
  };

  const removeAudio = () => {
    if (!audio.value) {
      console.error('No audio initialized', {
        url,
      });
      return;
    }

    audio.value.pause();
    audio.value = null;
    console.debug('Removing audio', {
      url,
      audio,
    });
  };

  return { playAudio, removeAudio };
}
```

Then you can load the audio by url (or path), and then have a ref to reference and play as needed without loading multiple times, and the ability to destroy it and clean up on component unmount? You could also modify the composable to use an imported audio file if needed.